### PR TITLE
Support Java 23 as latest version

### PR DESCRIPTION
**Pull request by akurtakov**
Needed so javac port can experiment with Java 23 features.

### CODE REVIEW

**Comment by akurtakov** [comment]
This is not suitable for upstreaming as nothing there supports Java 23 yet.

File: org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/impl/JavaFeature.java [hunk 1]
```diff
@@ -65,19 +65,19 @@ public enum JavaFeature {
...
-	STRING_TEMPLATES(ClassFileConstants.JDK22,
+	STRING_TEMPLATES(ClassFileConstants.JDK23,
 			Messages.bind(Messages.string_templates),
 			new char[][] {},
 			true),
-	STATEMENTS_BEFORE_SUPER(ClassFileConstants.JDK22,
+	STATEMENTS_BEFORE_SUPER(ClassFileConstants.JDK23,
 			Messages.bind(Messages.statements_before_super),
 			new char[][] {},
 			true),
```
Comment (fbricon) [inline org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/impl/JavaFeature.java]: I don't think you should change those values. Those features were introduced in 22

Comment (akurtakov) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/impl/JavaFeature.java]: I copied what was done in beta_java23 branch (https://github.com/eclipse-jdt/eclipse.jdt.core/commit/3e78c6e69a239a51d3d4e7c8c070ec488ce9fa89). Let me give it a try without this change.

Comment (akurtakov) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/impl/JavaFeature.java]: With and without the change JDT UI behaves the same when using JavaSE-23, unfortunately there is no way to enable preview features on JavaSE-22 if constant for support for Java 23 is there (somehow I believe it's UI limitation in JDT).

Comment (akurtakov) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/impl/JavaFeature.java]: And removed.

Comment (fbricon) [reply org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/impl/JavaFeature.java]: Supporting preview features for several compiler versions is not a priority IMO

**Comment by mickaelistria** [comment]
This has caused many failures on CI, including without Javac enabled.

**Comment by mickaelistria** [comment]
The failures can be seen at https://ci.eclipse.org/ls/job/jdt-core-incubator/job/dom-with-javac/67/testReport/ , the more important one are the ones prefixed with "Build and test with ECJ", many of them look like
```
configure.source is not updated.
----------- Expected ------------
source level should be in '1.1'...'1.8','9'...'23' (or '5.0'..'23.0'): 10
------------ but was ------------
source level should be in '1.1'...'1.8','9'...'22' (or '5.0'..'22.0'): 10
---------------------- ----------
 expected:<....1'...'1.8','9'...'2[3' (or '5.0'..'23].0'): 10> but was:<....1'...'1.8','9'...'2[2' (or '5.0'..'22].0'): 10>
	at org.eclipse.jdt.core.tests.junit.extension.TestCase.assertStringEquals(TestCase.java:267)
	at org.eclipse.jdt.core.tests.junit.extension.TestCase.assertEquals(TestCase.java:242)
	at org.eclipse.jdt.core.tests.compiler.regression.BatchCompilerTest.testBug573153(BatchCompilerTest.java:13300)
```

**Comment by akurtakov** [comment]
I see, adding the constants without ecj supporting them leads to that. I can't think of a (easy/acceptable) way to ease work on Java 23 and keep these working .
Feel free to revert if you guys think so. I more and more believe we will have to come up javac constants but this would not work well with JDT ui/debug.

**Comment by mickaelistria** [comment]
I'm not sure what's best here. Removing the constants makes us loose a very interesting feature we need to demo. Should the tests be tweaked to at least allow Java 23 to be present? Would more thing work if we don't set the "LATEST" to Java 23 (basically stick to Java 22 as latest, but allow Java 23 in the set anyway)?

**Comment by akurtakov** [comment]
LATEST is what jdt.ui uses to determine for which version to allow enabling preview features.

**Pull request merged**